### PR TITLE
Fixed leading spaces pushing text outside Label autowrap boundary

### LINKED ISSUE #59985: Leading spaces may push text outside `Label` autowrap boundary in 3.x
### Godot version

3.x (14856b1dbd7b428341f00f8819d53935d555b72c)

### System information

Arch Linux, GLES 3, Mesa Intel(R) UHD Graphics (CML GT2)

### Issue description

When autowrap is enabled, long leading spaces in `Label` may push text outside the boundary.

| Left Align | Center Align | Right Align | Fill |
| -- | -- | -- | -- |
| ksnip_20220407-221947 | ksnip_20220407-222001 | ksnip_20220407-222013 | ksnip_20220407-222024 |

Not in `master`. I believe it has been like this for quite a few versions.

### Steps to reproduce

Create a Label, enable Auto Wrap, put lots of spaces before actual text.

### Minimal reproduction project

[LabelSpace.zip](https://github.com/godotengine/godot/files/8443780/LabelSpace.zip)

### PR DESCRIPTION
**Pull request by ZheisterCoding**
Fixed #59985 leading spaces autowrap boundary problem

Bug problem:
No condition for when the first character of the label is a space character

Fix:
Added an IF condition for when the first character or the entered new line is a space character
The autowrap boundary treat this space character as another dummy word in the WordCache linked list and proceed to function normally
Added a local boolean variable `prev_insert_newline` to address leading space in newline issue

_*note: last pull request has rebase problem_

https://user-images.githubusercontent.com/50128969/163237340-3d1f745b-c00f-4600-9a3e-ff5116bf2703.mp4

@timothyqiu

### CODE REVIEW

**Comment by ZheisterCoding** [comment]
@timothyqiu I think the rebase and the commits should be correct now :)

File: scene/gui/label.cpp [hunk 1]
```diff
@@ -447,12 +452,19 @@ void Label::regenerate_word_cache() {
...
 			if (current == '\n') {
 				insert_newline = true;
+				prev_insert_newline = true;
 			} else if (current != ' ') {
 				total_char_cache++;
 			}
 
 			if (i < xl_text.length() && xl_text[i] == ' ') {
-				if (line_width > 0 || last == nullptr || last->char_pos != WordCache::CHAR_WRAPLINE) {
+				if (i == 0 || prev_insert_newline) {
```
Comment (timothyqiu) [inline scene/gui/label.cpp:461]: Having read your last description about how these variables work again. I think `prev_insert_newline` here is equivalent to checking `line_width == 0`. What do you think?

Comment (ZheisterCoding) [reply scene/gui/label.cpp]: @timothyqiu yes, that's true! I didn't think of that. I've implemented the fix without the `prev_insert_newline` variable and the codebase fix now looks alot cleaner and simple. I've pushed a new commit with your suggestion, when a newline is made, i set `line_width` back to 0. The fix works better than the last fix with the local variable :) thanks!

File: scene/gui/label.cpp [hunk 2]
```diff
@@ -447,12 +447,19 @@ void Label::regenerate_word_cache() {
 
 			if (current == '\n') {
 				insert_newline = true;
+				line_width = 0;
```
Comment (timothyqiu) [inline scene/gui/label.cpp:450]: I think you don't have to set `line_width` to 0.

`current_word_size` is set to 0 when `\n` is encountered:

https://github.com/godotengine/godot/blob/e1eb3c2fb1bcfb70204cb1d431720e316f544da1/scene/gui/label.cpp#L428

And `line_width` will be set to `current_word_size` when the `WordCache` for this newline is created:

https://github.com/godotengine/godot/blob/e1eb3c2fb1bcfb70204cb1d431720e316f544da1/scene/gui/label.cpp#L510

Comment (ZheisterCoding) [reply scene/gui/label.cpp]: @timothyqiu 
Alright, i've removed that statement, and the result of the fix did not change. :)

**Review by timothyqiu — APPROVED** [review]
LGTM. Please squash your commits. See instructions in [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase).

If you're not sure, I think you can first rebase your branch onto your _local_ `3.x` branch. If everything works fine, then when you do the _interactive_ rebase onto your local `3.x` branch, you'll see and only see your 9 commits listed in the editor.

**Comment by ZheisterCoding** [comment]
> LGTM. Please squash your commits. See instructions in [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase).
> 
> If you're not sure, I think you can first rebase your branch onto your _local_ `3.x` branch. If everything works fine, then when you do the _interactive_ rebase onto your local `3.x` branch, you'll see and only see your 9 commits listed in the editor.
@timothyqiu 
I think I might have to practice on rebasing first :') the merging conflicts problem keep appearing, I'm going to create a new pull request. Sorry for the extra work, sometimes interactive rebase works sometimes it doesnt for some reason.

This is the new pull request #60233

**Pull request closed without merging**